cloudify_azure/resources/deployment.py

```diff
@@ -163,3 +148,130 @@ def delete(ctx, **_):
+    ctx.logger.debug("initial_resources: {}".format(initial_resources))
+    actual_resources = resource_group.list_resources(resource_group_name)
+
+    deployment = Deployment(azure_config, ctx.logger)
+    properties, params = get_properties_and_formated_params(ctx, **kwargs)
+    template = ctx.instance.runtime_properties.get('template') or get_template(
+        ctx, properties)
+    # some resources are nested like subnets so they do not appear in
+    # actual_resources. We will search them using the what-if result.
+
+    what_if_res = execute_what_if(deployment,
+                                  resource_group_name,
+                                  deployment_name,
+                                  template,
+                                  params)
+    calculate_state(ctx, initial_resources, actual_resources, what_if_res)
+
+
+def get_properties_and_formated_params(ctx, **kwargs):
+    properties = {}
+    properties.update(ctx.node.properties)
+    properties.update(kwargs)
+    params = format_params(properties.get('params', {}))
+    return properties, params
+
+
+def get_resource_group_name_deployment_name_and_api_version(ctx):
+    deployment_name = utils.get_resource_name(ctx)
+    resource_group_name = ctx.node.properties.get(
+        'resource_group_name', deployment_name)
+    api_version = \
+        ctx.node.properties.get('api_version', constants.API_VER_RESOURCES)
+    return deployment_name, resource_group_name, api_version
+
+
+def execute_what_if(deployment,
+                    resource_group_name,
+                    deployment_name,
+                    template,
+                    params):
+    what_if_properties = {
+        'mode': DeploymentMode.incremental,
+        'template': template,
+        'parameters': params
+    }
+    try:
+        return deployment.what_if(resource_group_name,
+                                  deployment_name,
+                                  what_if_properties)
+    except CloudError:
+        raise cfy_exc.NonRecoverableError(
+            "What if operation on deployment {dep_name} failed. Can't "
+            "calculate accurate state.".format(dep_name=deployment_name))
+
+
+def calculate_state(ctx, initial_resources, actual_resources, what_if_res):
+    """
+    Create a list of live resources of the deployment.
+    Save this list in the state runtime property.
+    :param initial_resources: list of resources id`s created by the deployment.
+    :param actual_resources: list of resources id`s that exists in resource
+    group.
+    :param: result of what-if operation in order to indicate resources that
+    not appear in actual_resources.
+    """
+    state = []
+    initial_ids = [resource['id'] for resource in initial_resources]
+    actual_ids = [resource['id'] for resource in actual_resources]
+    for resource_id in initial_ids:
+        if resource_id not in actual_ids and not\
+                check_if_resource_alive_in_what_if_result(resource_id,
```

Comment:
Maybe the condition `if resource_id not in actual_ids` can be inside `check_if_resource_alive_in_what_if_result`?

Reply:
I think it's better as is because this condition doesn't belong to the search operation in what-if result.